### DISCUSSION TITLE
Does Lettuce implement L0 and L1 cache?

### DISCUSSION THREAD

**Discussion by jianwenl** [redis/lettuce#2049] [Q&A] [answered]
My cache use case needs L0 and L1 cache. Does anyone know if lettuce have L0 and L1 cache implementation, respectively ?  Thanks!

**Comment by mp911de** [comment]
What is a L0 cache in the context of a Redis client? L0/L1/… applies in that terminology to CPUs.

**Comment by jianwenl** [comment]
Sorry for the confusion. Both L0 and L1 cache refer to the client side cache. However, L0 cache resides on the JVM heap and can cache object without serialization directly, while L1 cache have to store serialized bytes.

**Comment by mp911de** [comment, marked as answer]
Lettuce ships with support for client-side caching, take a look at `io.lettuce.core.support.caching.ClientSideCaching` that provides a caching facade and registers invalidation listeners.

**Comment by jianwenl** [comment]
Thanks. Yes, I did check the `ClientSideCaching` doc, but I am not sure if this local cache implementation actually stores value in JVM heap (not requiring serialization) or not (requiring serialization)). Could you help me confirm this?

**Reply by mp911de** [reply]
The idea of client-side caching is to keep Lettuce as much as possible out of the actual cache. We provide accessors and a default implementation for caches that implement the Map interface. An example implementation can leverage a HashMap or EhCache. 

That being said, it serialization etc all depends on which backing cache you plug in.

**Reply by jianwenl** [reply]
Thanks. Let me rephrase my question, could lettuce `ClientSideCaching` store KV pairs as object to JVM heap without serialization? Sorry for the confusion of my question.

**Reply by mp911de** [reply]
Since Lettuce doesn't cache itself any objects (except if you use the `HashMap`-implementation), there's no serialization involved.

**Reply by jianwenl** [reply]
In summary, I can pass whatever things that implements a map interface to Lettuce client-side cache thanks to the design "idea of client-side caching is to keep Lettuce as much as possible out of the actual cache.". Thanks for reply, I will close this issue.
